lib/active_model/serializer/adapter/json.rb

```diff
@@ -47,6 +23,147 @@ def fragment_cache(cached_hash, non_cached_hash)
+              serialized_relationship,
+              serializer,
+              nested_associations_to_include)
+          end
+
+          serialized_relationship
+        end
+
+        # add a many relationship
+        def add_relationships(serializer, options)
+          serialize_array(serializer, options) do |serialized_item, item_serializer|
+            nested_associations_to_include = options[:include]
+
+            if nested_associations_to_include.present?
+              serialized_item = add_resource_relationships(
+                serialized_item,
+                item_serializer,
+                nested_associations_to_include)
+            end
+
+            serialized_item
+          end
+        end
+
+
+        def serialize_array_without_root(serializer, options)
+          serializer.map { |s| FlattenJson.new(s).serializable_hash(options) }
+        end
+
+        # a virtual value is something that doesn't need a serializer,
+        # such as a ruby array, or any other raw value
+        def serialized_or_virtual_of(serializer, options)
+          if serializer && serializer.object
+            resource_object_for(serializer)
+          elsif options[:virtual_value]
+            options[:virtual_value]
+          end
+        end
+
+        def serialize_array(serializer, options)
+          serializer.map do |item|
+            serialized_item = resource_object_for(item)
+            serialized_item = yield(serialized_item, item) if block_given?
+            serialized_item
+          end
+        end
+
+        def resource_object_for(serializer)
+          cache_check(serializer) do
+            serializer.attributes(serializer.options)
+          end
+        end
+
+        # converts the include hash to a standard format
+        # for constrained serialization recursion
+        #
+        # converts
+        #  [:author, :comments => [:author]] to
+        #  {:author => [], :comments => [:author]}
+        #
+        # and
+        #  [:author, :comments => {:author => :bio}, :posts => [:comments]] to
+        #  {:author => [], :comments => {:author => :bio}, :posts => [:comments]}
+        #
+        # The data passed in to this method should be an array where the last
+        # parameter is a hash
+        #
+        # the point of this method is to normalize the include
+        # options for the child relationships.
+        # if a sub inclusion is still an array after this method,
+        # it will get converted during the next iteration
+        def include_array_to_hash(include_array)
+          # still don't trust input
+          # but this also allows
+          #  include: :author syntax
+          include_array = Array[*include_array].compact
+
+          result = {}
+
+          hashes = include_array.select{|a| a.is_a?(Hash)}
+          non_hashes = include_array - hashes
+
+          hashes += non_hashes.map{ |association_name| { association_name => [] } }
```

Comment:
this seems more like a function than a (private) property of the JSON Adapter.. What would you think about adding a 

```ruby
module ActiveModel::Serializer::Utils
  module_function
  
        def include_array_to_hash(include_array)
          # still don't trust input
          # but this also allows
          #  include: :author syntax
          include_array = Array[*include_array].compact

          result = {}

          hashes = include_array.select{|a| a.is_a?(Hash)}
          non_hashes = include_array - hashes

          hashes += non_hashes.map{ |association_name| { association_name => [] } }

          # now merge all the hashes
          hashes.each{|hash| result.merge!(hash) }

          result
        end
end
```

and then `Utils.include_array_to_hash` or mix it in